lib/bambora/client.rb

```diff
@@ -1,47 +1,101 @@
+    #     c.batch_api_key = ENV.fetch('BAMBORA_BATCH_API_KEY')
+    #     c.reporting_api_key = ENV.fetch('BAMBORA_REPORTING_API_KEY')
+    #   end
+    #
+    # @param options[base_url] [String] Bambora Base URL
+    # @param options[merchant_id] [String] The Merchant ID for this request.
+    # @param options[sub_merchant_id] [String] The Sub-Merchant ID for this request.
+    # @param options[profiles_api_key] [String] The Profiles API Key for this request.
+    # @param options[payments_api_key] [String] The Payments API Key for this request.
+    # @param options[batch_api_key] [String] The Batch API Key for this request.
+    # @param options[reporting_api_key] [String] The Reporting API Key for this request.
     def initialize(options = {})
-      unless options[:version].nil?
-        raise Bambora::Error, 'Only V1 endpoints are supported at this time.' if options[:version].upcase != 'V1'
+      if !options[:version].nil? && options[:version].upcase != 'V1'
+        raise Bambora::Client::Error, 'Only V1 endpoints are supported at this time.'
       end
 
       options.each do |key, value|
         instance_variable_set("@#{key}", value)
       end
+
       yield(self) if block_given?
     end
 
-    protected
-
-    def get(path:, params:, headers:)
-      connection.get(path, params, headers)
+    # Retrieve a client to make requests the Profiles endpoints.
+    #
+    # @example
+    #   profiles = client.profiles
+    #   profiles.delete(customer_code: '02355E2e58Bf488EAB4EaFAD7083dB6A')
+    #
+    # @return [Bambora::V1::ProfileResource]
+    def profiles
+      @profiles ||= Bambora::V1::ProfileResource.new(client: json_client, api_key: profiles_api_key)
     end
 
-    def post(path:, body:, headers:)
-      connection.post(path, body, headers)
+    # Retrieve a client to make requests the Payments endpoints.
+    #
+    # @example
+    #   payments = client.profiles
+    #   payments.create(
+    #     {
+    #       amount: 50,
+    #       payment_method: 'card',
+    #       card: {
+    #         name: 'Hup Podling',
+    #         number: '4504481742333',
+    #         expiry_month: '12',
+    #         expiry_year: '20',
+    #         cvd: '123',
+    #       },
+    #     },
+    #   )
+    #
+    # @return [Bambora::V1::PaymentResource]
+    def payments
+      @payments ||= Bambora::V1::PaymentsResource.new(client: json_client, api_key: payments_api_key)
     end
 
-    def delete(path:, headers:)
-      connection.delete(path) do |req|
-        req.headers = headers
-      end
-    end
+    def batch_payment_reports; end
 
-    def connection
-      @connection ||= Faraday.new(url: base_url) do |f|
-        f.adapter :excon
-      end
-    end
+    def batch_payments; end
+
+    private
 
-    def build_headers(api_key)
-      Bambora::Headers.build(api_key: api_key, merchant_id: merchant_id, sub_merchant_id: sub_merchant_id)
+    def json_client
+      @json_client ||= Bambora::JSONClient.new(
```

Comment:
I don't think any of this would have been possible without moving the API keys to the `*Resource` classes, so that was a great change!